agent/resources/resources_linux.go

```diff
@@ -0,0 +1,120 @@
+// +build linux
+
+// Copyright 2017 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License"). You may
+// not use this file except in compliance with the License. A copy of the
+// License is located at
+//
+//	http://aws.amazon.com/apache2.0/
+//
+// or in the "license" file accompanying this file. This file is distributed
+// on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+// express or implied. See the License for the specific language governing
+// permissions and limitations under the License.
+
+package resources
+
+import (
+	"github.com/aws/amazon-ecs-agent/agent/api"
+	"github.com/aws/amazon-ecs-agent/agent/config"
+	"github.com/aws/amazon-ecs-agent/agent/resources/cgroup"
+	"github.com/cihub/seelog"
+	"github.com/pkg/errors"
+)
+
+// cgroupWrapper implements the Resource interface
+type cgroupWrapper struct {
+	control cgroup.Control
+}
+
+// New is used to return an object that implements the Resource interface
+func New() Resource {
+	return newResources(cgroup.New())
+}
+
+func newResources(control cgroup.Control) Resource {
+	return &cgroupWrapper{
+		control: control,
+	}
+}
+
+// Init is used to initialize the resource
+func (c *cgroupWrapper) Init() error {
+	return c.cgroupInit()
+}
+
+// Setup sets up the resource
+func (c *cgroupWrapper) Setup(task *api.Task) error {
+	return c.setupCgroup(task)
+}
+
+// Cleanup removes the resource
+func (c *cgroupWrapper) Cleanup(task *api.Task) error {
+	return c.cleanupCgroup(task)
+}
+
+// cgroupInit is used to create the root '/ecs/ cgroup
+func (c *cgroupWrapper) cgroupInit() error {
+	if c.control.Exists(config.DefaultTaskCgroupPrefix) {
+		seelog.Debugf("Cgroup at %s already exists, skipping creation", config.DefaultTaskCgroupPrefix)
+		return nil
+	}
+	return c.control.Init()
+}
+
+// setupCgroup is used to create the task cgroup
+func (c *cgroupWrapper) setupCgroup(task *api.Task) error {
+	cgroupRoot, err := task.BuildCgroupRoot()
+	if err != nil {
+		return errors.Wrapf(err, "resource: setup cgroup: unable to determine cgroup root for task: %s", task.Arn)
+	}
+
+	seelog.Debugf("Setting up cgroup at: %s for task: %s", cgroupRoot, task.Arn)
+
+	if c.control.Exists(cgroupRoot) {
+		seelog.Debugf("Cgroup at %s already exists, skipping creation", cgroupRoot)
+		return nil
+	}
+
+	linuxResourceSpec, err := task.BuildLinuxResourceSpec()
+	if err != nil {
+		return errors.Wrapf(err, "resource: setup cgroup: unable to build resource spec for task: %s", task.Arn)
+	}
+
+	cgroupSpec := cgroup.Spec{
+		Root:  cgroupRoot,
+		Specs: &linuxResourceSpec,
+	}
+
+	cgrp, err := c.control.Create(&cgroupSpec)
+	if err != nil {
+		return errors.Wrapf(err, "resource: setup cgroup: unable to create cgroup at %s for task: %s", cgroupRoot, task.Arn)
+	}
+
+	// NOTE: This should be impossible
+	// cgrp is an interface and it should never be nil when err == nil
+	if cgrp == nil {
```

Comment:
Please get rid of this if this is impossible (which it seems like from https://github.com/containerd/cgroups/blob/master/cgroup.go#L74).

Reply:
I still hold my case that having code to check for impossible conditions is fair. 
Isn't it better to fail to start a task rather than making the agent panic.  

I can remove it if you feel strongly about it.

Reply:
I'd rather panic if we ever hit an impossible condition as that'd mean something else has gone horribly wrong as well. This is dead code now for all practical purposes and I'd rather not have that in code.